# Exploratory Data Analysis

Databricks was the chosen environment to host all compute related activities so that the project is scalable for big data. We start by configuring a compute cluster and setting up this Jupyter environment with a spark session.

Import the necessary libraries...

In [ ]:
from pyspark.sql import SparkSession
from pyspark.sql import Row

Create spark session...

In [ ]:
spark = SparkSession.builder.appName('Exploratory_Data_Analysis').getOrCreate()

Upload data to `/FileStore/tables/` and run `dbutils.fs.ls("/FileStore/tables/")` to list all available files...

In [ ]:
# dbutils.fs.ls("/FileStore/tables/") ## Returns a list of all files in the Databricks filesystem (commented out for privacy)

Collect the data...

In [ ]:
class Dataset:
    def __init__(self, file_path:str) -> None:
        self.name = file_path.split('/')[-1]
        self.spark_df = spark.read.csv(file_path, inferSchema=True, header=True)
    @property
    def num_rows(self) -> int:
        return self.spark_df.count()
    @property
    def num_cols(self) -> int:
        return len(self.spark_df.columns)
    @property
    def columns(self) -> list[str]:
        return self.spark_df.columns

folder_path = '/FileStore/tables/'
file_names = [
    'test.csv'
    , 'train.csv'
    , 'Customer_Churn_Records.csv'
    , 'Bank_Customer_Churn_Prediction.csv'
    , 'Churn_Modeling.csv'
    , 'Churn_Modelling.csv'
    , 'Churn_Modelling-1.csv'
    , 'churn.csv'
]
file_paths = [folder_path+f for f in file_names]
df_lst = [Dataset(f) for f in file_paths]

In [ ]:
print(f'There are {len(df_lst)} spark dataframes.')

There are 8 spark dataframes.


Analysing the various sizes of the datasets...

In [ ]:
shape_df = spark.createDataFrame(
    [Row(
        df_index=i
        , number_of_columns=sdf.num_cols
        , number_of_rows=sdf.num_rows
        , file_name=sdf.name
    ) for i,sdf in enumerate(df_lst)]
)

In [ ]:
shape_df.show(truncate=False)

+--------+-----------------+--------------+----------------------------------+
|df_index|number_of_columns|number_of_rows|file_name                         |
+--------+-----------------+--------------+----------------------------------+
|0       |13               |110023        |test.csv                          |
|1       |14               |165034        |train.csv                         |
|2       |18               |10000         |Customer_Churn_Records.csv        |
|3       |12               |10000         |Bank_Customer_Churn_Prediction.csv|
|4       |14               |10000         |Churn_Modeling.csv                |
|5       |14               |10000         |Churn_Modelling.csv               |
|6       |14               |10002         |Churn_Modelling-1.csv             |
|7       |14               |10000         |churn.csv                         |
+--------+-----------------+--------------+----------------------------------+



In [ ]:
max_num_cols = shape_df.agg({'number_of_columns':'max'}).collect()[0][0]
print(f'There are up to {max_num_cols} columns.')

There are up to 18 columns.


Example schema:

In [ ]:
df_lst[3].spark_df.printSchema()

root
 |-- customer_id: integer (nullable = true)
 |-- credit_score: integer (nullable = true)
 |-- country: string (nullable = true)
 |-- gender: string (nullable = true)
 |-- age: integer (nullable = true)
 |-- tenure: integer (nullable = true)
 |-- balance: double (nullable = true)
 |-- products_number: integer (nullable = true)
 |-- credit_card: integer (nullable = true)
 |-- active_member: integer (nullable = true)
 |-- estimated_salary: double (nullable = true)
 |-- churn: integer (nullable = true)



Analysing the column names...

In [ ]:
col_tuples = [[c for c in sdf.columns] for sdf in df_lst]
for t in col_tuples:
    while len(t) < shape_df.agg({'number_of_columns':'max'}).collect()[0][0]:
        t.append(None)
col_tuples = [[str(i)]+t for i,t in enumerate(col_tuples[:])]
col_tuples = [tuple(t) for t in col_tuples[:]]

column_pandas_df = spark.createDataFrame(
    [Row(*t) for t in col_tuples]
)

In [ ]:
column_pandas_df.show(truncate=True)

+---+-----------+------------+-------+-----------+---------+------+-------+---------------+-----------+-------------+----------------+--------------+---------------+------+--------+------------------+---------+------------+
| _1|         _2|          _3|     _4|         _5|       _6|    _7|     _8|             _9|        _10|          _11|             _12|           _13|            _14|   _15|     _16|               _17|      _18|         _19|
+---+-----------+------------+-------+-----------+---------+------+-------+---------------+-----------+-------------+----------------+--------------+---------------+------+--------+------------------+---------+------------+
|  0|         id|  CustomerId|Surname|CreditScore|Geography|Gender|    Age|         Tenure|    Balance|NumOfProducts|       HasCrCard|IsActiveMember|EstimatedSalary|  null|    null|              null|     null|        null|
|  1|         id|  CustomerId|Surname|CreditScore|Geography|Gender|    Age|         Tenure|    Balance|N

What looked good in Pandas does not look good as a Spark dataframe. Approaching representation in a different angle...

In [ ]:
column_melt_df = spark.createDataFrame(
    [Row(
        df_index=i
        , column_name=c
        , file_name=sdf.name
    ) for i,sdf in enumerate(df_lst) for c in sdf.columns]
)

In [ ]:
column_melt_df.count()

Out[13]: 113

In [ ]:
column_melt_df.show(113, truncate=False)

+--------+------------------+----------------------------------+
|df_index|column_name       |file_name                         |
+--------+------------------+----------------------------------+
|0       |id                |test.csv                          |
|0       |CustomerId        |test.csv                          |
|0       |Surname           |test.csv                          |
|0       |CreditScore       |test.csv                          |
|0       |Geography         |test.csv                          |
|0       |Gender            |test.csv                          |
|0       |Age               |test.csv                          |
|0       |Tenure            |test.csv                          |
|0       |Balance           |test.csv                          |
|0       |NumOfProducts     |test.csv                          |
|0       |HasCrCard         |test.csv                          |
|0       |IsActiveMember    |test.csv                          |
|0       |EstimatedSalary

In [ ]:
column_melt_df.select('column_name').dropDuplicates().count()

Out[15]: 31

In [ ]:
column_melt_df.select('column_name').dropDuplicates().show(31)

+------------------+
|       column_name|
+------------------+
|     NumOfProducts|
|           Balance|
|       CreditScore|
|               Age|
|        CustomerId|
|         HasCrCard|
|         Geography|
|            Tenure|
|           Surname|
|            Gender|
|    IsActiveMember|
|                id|
|   EstimatedSalary|
|         RowNumber|
|            Exited|
|          Complain|
|     active_member|
|      credit_score|
|           balance|
|         Card Type|
|      Point Earned|
|Satisfaction Score|
|       customer_id|
|            tenure|
|       credit_card|
|           country|
|               age|
|            gender|
|   products_number|
|  estimated_salary|
|             churn|
+------------------+



Given that there were up to 18 columns in `shape_df.show(truncate=False)`, 31 columns in `column_df.count()` seems a bit much. The line `column_df.select('column_name').dropDuplicates().show(31)` confirms this when we see duplicate column names such as "age" and "Age".

In [ ]:
column_melt_df.filter("column_name = 'Geography'").show(truncate=False)

+--------+-----------+--------------------------+
|df_index|column_name|file_name                 |
+--------+-----------+--------------------------+
|0       |Geography  |test.csv                  |
|1       |Geography  |train.csv                 |
|2       |Geography  |Customer_Churn_Records.csv|
|4       |Geography  |Churn_Modeling.csv        |
|5       |Geography  |Churn_Modelling.csv       |
|6       |Geography  |Churn_Modelling-1.csv     |
|7       |Geography  |churn.csv                 |
+--------+-----------+--------------------------+



A lot of the datasets appear to be duplicates. Selecting the first row from each dataset yields...

In [ ]:
for sdf in df_lst:
    print(sdf.spark_df.first().asDict().values())

dict_values([165034, 15773898, 'Lucchese', 586, 'France', 'Female', 23.0, 2, 0.0, 2, 0.0, 1.0, 160976.75])
dict_values([0, 15674932, 'Okwudilichukwu', 668, 'France', 'Male', 33.0, 3, 0.0, 2, 1.0, 0.0, 181449.97, 0])
dict_values([1, 15634602, 'Hargrave', 619, 'France', 'Female', 42, 2, 0.0, 1, 1, 1, 101348.88, 1, 1, 2, 'DIAMOND', 464])
dict_values([15634602, 619, 'France', 'Female', 42, 2, 0.0, 1, 1, 1, 101348.88, 1])
dict_values([1, 15634602, 'Hargrave', 619, 'France', 'Female', 42, 2, 0.0, 1, 1, 1, 101348.88, 1])
dict_values([1, 15634602, 'Hargrave', 619, 'France', 'Female', 42, 2, 0.0, 1, 1, 1, 101348.88, 1])
dict_values([1, 15634602, 'Hargrave', 619, 'France', 'Female', 42.0, 2, 0.0, 1, 1, 1, 101348.88, 1])
dict_values([1, 15634602, 'Hargrave', 619, 'France', 'Female', 42, 2, 0.0, 1, 1, 1, 101348.88, 1])


Setting aside `train.csv` and `test.csv`, because clearly this data has been shuffled for ML purposes... It appears that the rest of the data seems to all be the same data. Remember each dataset had approx 10,000 rows and it can be clearly seen from the first row of each dataset that they all represent the same female customer from France with credit score 619 etc.

In conclusion, there may only be 2 groups of data instead of 7 groups of data. Duplicate data may come in handy when validating data during the integration process in the upcoming data and ML pipelines...